mls-rs-uniffi/src/lib.rs

```diff
@@ -372,94 +388,64 @@ impl Client {
+    type Error = Error;
 
-impl RatchetTree {
-    // TODO(mgeisler): merge with #[uniffi::export] impl above when
-    // https://github.com/mozilla/uniffi-rs/issues/1074 is fixed.
-    /// Decode a ratched tree from its MLS encoding.
-    pub fn from_bytes(bytes: &[u8]) -> Result<Self, Error> {
-        let exported_tree = mls_rs::group::ExportedTree::from_bytes(bytes)?;
-        Ok(exported_tree.into())
+    fn try_from(exported_tree: mls_rs::group::ExportedTree<'static>) -> Result<Self, Error> {
+        let bytes = exported_tree.to_bytes()?;
+        Ok(Self { bytes })
     }
 }
 
-// TODO(mgeisler): remove this when associated functions are supported
-// by UniFFI: https://github.com/mozilla/uniffi-rs/issues/1074.
-#[uniffi::export]
-/// Decode a ratched tree from its MLS encoding.
-pub fn ratchet_tree_from_bytes(bytes: &[u8]) -> Result<RatchetTree, Error> {
-    RatchetTree::from_bytes(bytes)
-}
-
-#[derive(Clone, Debug, uniffi::Object)]
+#[derive(Clone, Debug, uniffi::Record)]
 pub struct CommitOutput {
-    inner: mls_rs::group::CommitOutput,
-}
-
-#[uniffi::export]
-impl CommitOutput {
     /// Commit message to send to other group members.
-    pub fn commit_message(&self) -> Message {
-        self.inner.commit_message.clone().into()
-    }
+    pub commit_message: Arc<Message>,
 
     /// Welcome message to send to new group members.
-    pub fn welcome_messages(&self) -> Vec<Arc<Message>> {
-        self.inner
-            .welcome_messages
-            .iter()
-            .map(|welcome_message| Arc::new(welcome_message.clone().into()))
-            .collect::<Vec<_>>()
-    }
+    pub welcome_messages: Vec<Arc<Message>>,
 
     /// Ratchet tree that can be sent out of band if the ratchet tree
     /// extension is not used.
-    pub fn ratchet_tree(&self) -> Option<Arc<RatchetTree>> {
-        self.inner
-            .ratchet_tree
-            .as_ref()
-            .map(|ratchet_tree| Arc::new(ratchet_tree.clone().into()))
-    }
+    pub ratchet_tree: Option<RatchetTree>,
 
     /// A group info that can be provided to new members in order to
     /// enable external commit functionality.
-    pub fn group_info(&self) -> Option<Arc<Message>> {
-        self.inner
-            .external_commit_group_info
-            .as_ref()
-            .map(|group_info| Arc::new(group_info.clone().into()))
-    }
-
+    pub group_info: Option<Arc<Message>>,
     // TODO(mgeisler): decide if we should expose unused_proposals()
     // as well.
 }
 
-impl From<mls_rs::group::CommitOutput> for CommitOutput {
-    fn from(inner: mls_rs::group::CommitOutput) -> Self {
-        Self { inner }
+impl TryFrom<mls_rs::group::CommitOutput> for CommitOutput {
+    type Error = Error;
+
+    fn try_from(commit_output: mls_rs::group::CommitOutput) -> Result<Self, Error> {
+        let commit_message = Arc::new(commit_output.commit_message.clone().into());
+        let welcome_messages = commit_output
+            .welcome_messages
+            .iter()
```

Comment:
Can `.into_iter()` be used to avoid cloning in `.map(...)`?

Reply:
Nice!